Add list len refinement

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56966 [JIT] Disable Complete Shape Inlining For Testing Purposes
* #56828 Add unary/binary ops necessary for mobilenet
* #56642 Only optimize while change has been made
* #56438 [JIT] integer value refinement
* #56437 Factor out isDominatedBy
* #55978 Add Value * == Value * peephole
* #55977 [JIT] [Easy] Factor out non tensor peephole
* **#55926 Add list len refinement**
* #55925 Add handling of symbolic shapes
* #54809 Initial Symbolic Shape Analysis

This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

Commit:
Add list len refinement

[ghstack-poisoned]

Commit:
Update on "Add list len refinement"




[ghstack-poisoned]

Commit:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators. 


[ghstack-poisoned]

Commit:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators. 


[ghstack-poisoned]

Commit:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators. 


[ghstack-poisoned]

Commit:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators. 


[ghstack-poisoned]

Commit:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

Commit:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

Commit:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+  };
+
+  c10::optional<int64_t> tryFindRefinement(Value* v) {
+    for (const auto& ref : active_refinements_) {
+      auto maybe_refinement = ref->find(v);
+      if (maybe_refinement != ref->end()) {
+        return maybe_refinement->second;
+      }
+    }
```
Is it possible to have several refinements for the same value? Here we only return the first one.

eellison replied:
It's not. For that to happen you would need two different contradicting refinements which would be unreachable code
```
if len(x) == 4:
    if len(x) == 5:
          # unreachable code 
```

ZolotukhinM commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+        }
+
+        // if either block has a constant bool output, e.g. `true` on the
+        // truee block, then for the `false` value we can take the false
```
Typo: truee

ZolotukhinM commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+      }
+
+      if (n->kind() == prim::If) {
+        IfView if_n(n);
```
Nit: might be worth factoring this out to a separate method.

eellison replied:
I'm doing this in a follow up PR to share logic between this and another pass so going to leave out of this one if that's fine.

ZolotukhinM reviewed (commented):
I think I get the gist of the code and it seems good, but I'd like to take another pass to understand it better. If you add more comments, it'd help me too :)

Commit:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

penguinwu commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+          continue;
+        }
+
+        // if either block has a constant bool output, e.g. `true` on the
```
This part of the code is the hardest to understand, maybe adding an example in the comment would help?

penguinwu commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+          if (auto maybe_bool = constant_as<bool>(true_v)) {
+            if (*maybe_bool) {
+              out = BoolRefinements::FalseRefinements(unionRefinements(
+                  info_[false_v].false_refine(), false_block_refinements));
```
From Line 258-259, it indicates `info_count[true_v] || info_count[false_v]`, so does `info_[false_v]` always exists under this condition? If so, please add some comments because it is not obvious.

eellison replied:
Yea, will do. We never associate refinements with constant values so yes that is the invariant.

penguinwu commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+                  info_[true_v].true_refine(), true_block_refinements));
+            }
+          }
+          if (info_.count(true_v) && info_.count(false_v)) {
```
I think I understand it now after several readings, but perhaps adding an example would help a lot.
```
if (len(x) == 4) {
   assert(len(y) == 5); 
} else {
  assert(len(y) == 5);
}
```
In this case, both true and false branches have refined length. So the refined length after the if-statement has to be the intersection of the two.

penguinwu commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+}
+
+// Represents the refinement information that can be carried on a boolean
+struct BoolRefinements {
```
I think adding some examples of `TrueRefinement`, `FalseRefinement`, and `BoolRefine` w/ both `true_refine` and `false_refine` would make it easier to understand.

eellison replied:
Happy for naming suggestions.... We're associating Refinements with the boolean's true value, and different Refinements for the boolean's false value.

ZolotukhinM replied:
One other slightly odd thing about `BoolRefinement` in comparison to `ListRefinement` is that the latter is a map from Value to some info we know about that value. But the former is itself a data we store in some map. I wonder if it would be more consistent if we called that outer map a `BoolRefinement`. Maybe what we have now is perfectly fine too, it's just something difficult to wrap one's head around.

eellison replied:
Not all refinements are associated with a boolean value. 
If you have something like
```
assert len(x) == 4
print(len(x))
```

After the assertion there are refinements present in the block that are not mapped to a particular value. So I'm not sure it makes sense to add a mapping of (BoolValue) to the core data structure, and would prefer the extra indirection of Value* -> pair(refinements) on the class itself.

ZolotukhinM replied:
Yeah, that's fine. What I was pointing out is that we use the word "refinement" in one situation to describe info about all values (list values), and in the other situation about a single value, and that seems inconsistent.

penguinwu commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+
+// To union, just take all refinements from both inputs. We do not need to worry
+// about len refinements disagreeing because a path like `if len(x) == 4 and
+// len(x) == 5` will never be taken
```
Adding an example on when UnionRefinements are used would be helpful for understanding.

penguinwu commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+// TODO: vector may be faster
+using ListRefinement = std::unordered_map<Value*, int64_t>;
+
+// The intersection of the refinements is the Value* which are in both
```
Adding an example of when intersectRefinements are applied would help understanding.

penguinwu reviewed (changes requested):
This code is really nice. My comments are mostly for understandability. I took me a couple of reads to understand most of it but it is not straightforward so I suggest to add some examples in the comments.

BTW, I do not seem to see the handling for `and`/`or`, are they all converted to `prim:If` nodes?

eellison commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
 
+// Refine from Value -> len of list
+// TODO: vector may be faster
+using ListRefinement = std::unordered_map<Value*, int64_t>;
```
Yes will do, that's correct.

eellison commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+  // A stack of active refinements, one for each block
+  std::vector<ListRefinement*> active_refinements_;
+  // A map from Boolean Value * -> associated refinements
+  std::unordered_map<Value*, BoolRefinements> info_;
```
Yea, I don't really know to call this :'( Maybe, `boolean_values_to_refinement_information_` ? Open to suggestions...

eellison reviewed (commented):
Thanks for comments, all. Going to try to add more examples and make it more readable.. Agreed its a little tricky as is.

eellison commented:
@penguinwu `and` and `or` get desugared to prim::Ifs

Commit:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

Commit:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

Commit:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -21,17 +22,365 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+          // true block, then for the `false` value we can take the false
+          // refinements present on the false block and from the other block
+          // output value bc if the output is false it had to have come from the
+          // false block. if len(a) == 5:
```
Nit: start example from a new line.

Nit2: the description goes about a case where we have a true on the true branch, but example illustrates false on false branch. It's probably better to use the same case.

Nit3: might be worth writing examples in torch-scripts, bc in python true_v and false_v are both "x", while in TS they would be explicitly different.

ZolotukhinM commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -21,17 +22,365 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+            }
+          } else if (auto maybe_bool = constant_as<bool>(false_v)) {
+            if (*maybe_bool) {
+              out = BooleanRefinementMapping::FalseRefinements(unionRefinements(
```
Nit: would be helpful to add an example to each of these cases to make it easier trace the logic.

ZolotukhinM commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -21,17 +22,365 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+                  boolean_value_refinements_[true_v].false_refine(),
+                  true_block_refinements));
+            } else {
+              out = BooleanRefinementMapping::TrueRefinements(unionRefinements(
+                  boolean_value_refinements_[true_v].true_refine(),
+                  true_block_refinements));
```
I'm not sure I understand logic of these cases. I tried to walk through some examples and my results disagree with the logic here IIUIC:

```
Case A:
x0 = if len(a) == 5:
     |   x1 = len(b) == 4
     | else:
     |   x2 = False

if x0 is False then (len(b) != 4 OR len(a) != 5) ==> no guarantees can be derived
if x0 is True  then (len(b) == 4 AND len(a) == 5)
x0_refinements = {true: {},
                  false: {b->4, a->5}}

Case B:
x0 = if len(a) == 5:
     |   x1 = len(b) == 4
     | else:
     |   x2 = True

if x0 is False then (len(a) == 5 AND len(b) != 4) ==> len(a) is guaranteed to be 5
if x0 is True  then (len(a) != 5 OR  len(b) == 4) ==> no guarantees can be derived
x0_refinements = {true: {a->5}, false: {}}

Case C:
x0 = if len(a) == 5:
     |   x1 = True
     | else:
     |   x2 = len(b) == 4

if x0 is False then (len(a) != 5 AND len(b) != 4) ==> no guarantees can be derived
if x0 is True  then (len(a) == 5 OR  len(b) == 4) ==> no guarantees can be derived
x0_refinements = {true: {}, false: {}}

Case D:
x0 = if len(a) == 5:
     |   x1 = False
     | else:
     |   x2 = len(b) == 4

if x0 is False then (len(a) == 5 OR len(b) != 4) ==> no guarantees can be derived
if x0 is True  then (len(a) != 5 OR len(b) == 4) ==> no guarantees can be derived
x0_refinements = {true: {}, false: {}}
```

Probably I still don't understand how it works - could you please point me to where I made a mistake in these equations?

eellison replied:
discussed offline

Commit:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

ZolotukhinM reviewed (approved):
Looks good to me, thanks for detailed explanations! :)

eellison committed:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

eellison committed:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750997).

eellison committed:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750997).

eellison committed:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750997).

eellison committed:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750997).

eellison committed:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750997).

eellison committed:
Update on "Add list len refinement"


This is necessary for code like conv2d where we wish to share a generic convolution shape function logic with that of conv2d but for conv2d always infer the output is dimension 4. I'm also hoping the refinement algorithm here could be refactored out and used to support refining tensor types from user annotations. i have a length comment explaining how this works, and the logic outside of data structures is pretty small and contained. Additionally, you might check out https://fb.quip.com/X7EVAdQ99Zzm for a very similar description of how to refine values based on comparison operators.

Differential Revision: [D27750997](https://our.internmc.facebook.com/intern/diff/D27750997)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750997).
